web/hooks/push.go

```diff
@@ -0,0 +1,286 @@
+
+	switch {
+	case repo.IsTestsRepo():
+		if !isDefaultBranch(payload) {
+			wh.logger.Debugf("Ignoring push event for non-default branch: %s", payload.GetRef())
+			return
+		}
+		// the push event is for the 'tests' repo, which means that we
+		// should update the course data (assignments) in the database
+		assignments.UpdateFromTestsRepo(wh.logger, wh.db, course)
+
+	case repo.IsUserRepo():
+		wh.logger.Debugf("Processing push event for user repo %s", payload.GetRepo().GetName())
+		if !isDefaultBranch(payload) {
+			wh.logger.Debugf("Ignoring push event for non-default branch: %s", payload.GetRef())
+			return
+		}
+		wh.updateLastActivityDate(repo.UserID, course.ID)
+		assignments := wh.extractAssignments(payload, course)
+		for _, assignment := range assignments {
+			if !assignment.IsGroupLab {
+				// only run non-group assignments
+				wh.runAssignmentTests(assignment, repo, course, payload)
+			} else {
+				wh.logger.Debugf("Ignoring assignment: %s, pushed to user repo: %s", assignment.GetName(), payload.GetRepo().GetName())
+			}
+		}
+
+	case repo.IsGroupRepo():
+		wh.logger.Debugf("Processing push event for group repo %s", payload.GetRepo().GetName())
+		jobOwner, _, err := wh.db.GetUserByCourse(course, payload.GetSender().GetLogin())
+		if err != nil {
+			wh.logger.Errorf("Failed to find user %s in course %s: %v", payload.GetSender().GetLogin(), course.GetName(), err)
+			return
+		}
+		wh.updateLastActivityDate(jobOwner.ID, course.ID)
+		assignments := wh.extractAssignments(payload, course)
+		for _, assignment := range assignments {
+			if assignment.IsGroupLab {
+				// only run group assignments
+				results := wh.runAssignmentTests(assignment, repo, course, payload)
+				if !isDefaultBranch(payload) {
+					// Attempt to find the pull request for the branch, if it exists,
+					// and then assign reviewers to it, if the branch task score is higher than the assignment score limit
+					wh.handlePullRequestPush(payload, results, assignment, course, repo)
+				}
+			} else {
+				wh.logger.Debugf("Ignoring assignment: %s, pushed to group repo: %s", assignment.GetName(), payload.GetRepo().GetName())
+			}
+		}
+	default:
+		wh.logger.Debug("Nothing to do for this push event")
+	}
+}
+
+// handlePullRequestPush attempts to find a pull requested associated with the non-default branch.
+// If successfull, it then finds the relevant task, and uses it to receive the relevant task score.
+// If a passing score is reached, it assigns reviewers to the pull request.
+// It also uses the test results and task to generate a feedback comment for the pull request.
+func (wh GitHubWebHook) handlePullRequestPush(payload *github.PushEvent, results *score.Results, assignment *pb.Assignment, course *pb.Course, repo *pb.Repository) {
+	wh.logger.Debugf("Attempting to find pull request for ref: %s, in repository: %s",
+		payload.GetRef(), payload.GetRepo().GetFullName())
+
+	pullRequest, err := wh.db.GetPullRequest(&pb.PullRequest{
```

Comment:
Suggest to move Lines 105-131 to a separate method that returns `pullRequest, taskName`. The rest of this method (after 131) only needs the `task.LocalName()`. See related comment in `pull_request.go:CreateFeedbackComment`.